### DISCUSSION TITLE
CSharpCAD vs JSCAD Performance Numbers

### DISCUSSION THREAD

**Discussion by briansturgill** [jscad/OpenJSCAD.org#1101] [General]
I translated the JSCAD performance tests to CSharpCAD and ran timing tests on the same system.
Attached are three items, jscad_perf.txt, cscad_perf.txt, and myperf.js.
I ran the myperf.js command as: node myperf.js
The suggested node flags for GC in the Performance Tests README are no longer valid.
I used node version: v12.22.9 (current under Ubuntu MATE).
Myperf.js is a rewrite of the the performanceV3.js -- The microtime module used was either faulty, or systems have gotten too fast to use it.
Also, I didn't like it scaling the numbers down to 0.000000.....  Hard to read, I had to keep counting zeros.
The myperf.js output is in the number of milliseconds it took to run the test 1000 times.
The same sort of mechanism was used in the C# test harness.

The tests were run on CSCAD and JSCAD with NEITHER OF THEM having the optimizations to Poly3.MeasureBoundingBox/Sphere and Math.hypot that I've brought up in new issues recently.

SUMMARY:
As C# is a compiled language, its not surprising that CSharpCAD runs everything, generally speaking, 2x+ faster than JSCAD.
JSCAD runs things faster only in very short running code... I suspect the JSCAD Jitter is lighter weight than the C# Jitter.
For CPU intensive items, CSCAD runs booleans operations about 1.8 times faster than JSCAD... however once the discussed optimizations to Poly3.measureBoundBox... etc. are made, I'm expecting that difference might only be 1.5 times faster as JSCAD is particularly hindered here by the slowness of Math.hypot under nodeJS.
Otherwise, it's CSCAD all the way with extrusions and hulls runnning 4x+ faster (a large extrudeRectangular is 10x faster!).

## myperf.js
```javascript
const tests = require('./testsV2');

let entries = Object.entries(tests)
entries.forEach((entry) => {
  // setup the performance test
  //if (entry && (typeof(entry[1]) === 'function')) {
  if (entry) {
    let item = entry[0]
    let test = entry[1]

    // obtain the details of the test
    let name = test.name
    let api  = test.api
    let div  = test.div
    let performance = test.func

    //console.log('Test: '+name+' ('+api+','+div+')')

    let samples = 0
    let maxsamples = 1000

    if (api === 'setup') maxsamples = 1

		let tag = api+','+div+','+maxsamples
    if (maxsamples > 1) {
			console.time(tag)
		}
    while (samples < maxsamples) {
      performance()
      samples++
    }

    if (maxsamples > 1) {
      console.timeEnd(tag)
    }
  }
})
```

[cscad_perf.txt](https://github.com/jscad/OpenJSCAD.org/files/8817660/cscad_perf.txt)

[jscad_perf.txt](https://github.com/jscad/OpenJSCAD.org/files/8817667/jscad_perf.txt)

**Comment by briansturgill** [comment]
Are there tests comparing OpenSCAD and JSCAD performance?

**Comment by platypii** [comment]
Very interesting data, thanks @briansturgill. That's not terrible performance, considering it's JavaScript. But I'm not done optimizing yet :smiley:

**Reply by briansturgill** [reply]
Yes, I concur, writing a CAD package in JavaScript is a difficult task. You guys are doing an amazing job.

**Comment by platypii** [comment]
Node has gotten a fair bit faster since version 12, I've seen our tests get faster with newer node version. I'd be curious how the numbers compare on a newer node version, and the `no-hypot` branch.

**Reply by briansturgill** [reply]
I grabbed node-18.2.0 and am running it now. This would be the exact same jscad as the first test.
I'll run a second set on both after performance work has been done on both sides.
Note, you need to do more than just fix hypot... the poly3.measureBoundingSphere and all that it calls is 57% of union test CPU usage.
That needs to be as optimized as possible. It's the best bet for good boolean performance, which is probably the most important part of CSG-based CAD.

**Reply by z3dev** [reply]
> Note, you need to do more than just fix hypot... the poly3.measureBoundingSphere and all that it calls is 57% of union test CPU usage.

concur. Those extra memory allocations in measureBoundingBox / sphere are taking a lot of time. The booleans probably need a special version that reuses an array, and faster loops.

A special hypot in the booleans would also be fine.

**Reply by briansturgill** [reply]
So far, there is maybe 20% speed up in many areas and surprisingly, the 10x slow down in extrudeRectangle is now less than 3x slower.
Alas, not doing well on intersect though. Actually slower in more complex intersections.

**Reply by platypii** [reply]
I have another work-in-progress branch to improve the booleans even more by caching `poly3.measureBoundingSphere`. You were right @briansturgill there is another 20-30% of available performance increase in the booleans even AFTER the hypot change. :star_struck:

**Comment by briansturgill** [comment]
OK, the test have finished under node 18.2.0 -- generally somewhat faster.
Boolean performance disappointing... intersect and subtract lost some ground, union gained a little.
Note node.js changed the output of console.timeEnd... you'll need a calculator if the time goes into minutes.

[jscad_perf-node18.2.0.txt](https://github.com/jscad/OpenJSCAD.org/files/8819911/jscad_perf-node18.2.0.txt)
